<h1>Activity 1 - PALMER PENGUINS DATA ANALYTICS PROJECT </h1>
<hr>
<h3>Analyst : Lovely Claire Cabana<h3>

In [73]:
#import libraries
import numpy as np
import pandas as pd 
import warnings
warnings.filterwarnings('ignore')

<h2>Converting Dataset(s) into DataFrame</h2>

<h2>REQUIREMENTS</h2>
<ul>
    <li>
        Name of the DataFrame
    </li>
    <li>
        File Path
    </li>
</ul>

In [74]:
penguins = pd.read_csv('Datasets\\penguins.csv')

In [75]:
penguins

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen            39.1           18.7   
1        2     Adelie  Torgersen            39.5           17.4   
2        3     Adelie  Torgersen            40.3           18.0   
3        4     Adelie  Torgersen             NaN            NaN   
4        5     Adelie  Torgersen            36.7           19.3   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream            55.8           19.8   
340    341  Chinstrap      Dream            43.5           18.1   
341    342  Chinstrap      Dream            49.6           18.2   
342    343  Chinstrap      Dream            50.8           19.0   
343    344  Chinstrap      Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
0                181.0       3750.0    male  2007  
1                186.0       3800.0  female  2007  
2                195.0       3250.0  female  2007  
3                  NaN          NaN     NaN  2007  
4                193.0       3450.0  female  2007  
..                 ...          ...     ...   ...  
339              207.0       4000.0    male  2009  
340              202.0       3400.0  female  2009  
341              193.0       3775.0    male  2009  
342              210.0       4100.0    male  2009  
343              198.0       3775.0  female  2009  

[344 rows x 9 columns]

<h2>DESCRIPTIVE ANALYTICS</h2>

In [76]:
penguins.describe()

rowid  bill_length_mm  bill_depth_mm  flipper_length_mm  \
count  344.000000      342.000000     342.000000         342.000000   
mean   172.500000       43.921930      17.151170         200.915205   
std     99.448479        5.459584       1.974793          14.061714   
min      1.000000       32.100000      13.100000         172.000000   
25%     86.750000       39.225000      15.600000         190.000000   
50%    172.500000       44.450000      17.300000         197.000000   
75%    258.250000       48.500000      18.700000         213.000000   
max    344.000000       59.600000      21.500000         231.000000   

       body_mass_g         year  
count   342.000000   344.000000  
mean   4201.754386  2008.029070  
std     801.954536     0.818356  
min    2700.000000  2007.000000  
25%    3550.000000  2007.000000  
50%    4050.000000  2008.000000  
75%    4750.000000  2009.000000  
max    6300.000000  2009.000000

In [77]:
penguins.describe().T

count         mean         std     min       25%      50%  \
rowid              344.0   172.500000   99.448479     1.0    86.750   172.50   
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                       75%     max  
rowid               258.25   344.0  
bill_length_mm       48.50    59.6  
bill_depth_mm        18.70    21.5  
flipper_length_mm   213.00   231.0  
body_mass_g        4750.00  6300.0  
year               2009.00  2009.0

<h2>CHECKING FOR NULL VALUES</h2>

In [78]:
penguins.isnull()

rowid  species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    False    False   False           False          False              False   
1    False    False   False           False          False              False   
2    False    False   False           False          False              False   
3    False    False   False            True           True               True   
4    False    False   False           False          False              False   
..     ...      ...     ...             ...            ...                ...   
339  False    False   False           False          False              False   
340  False    False   False           False          False              False   
341  False    False   False           False          False              False   
342  False    False   False           False          False              False   
343  False    False   False           False          False              False   

     body_mass_g    sex   year  
0          False  False  False  
1          False  False  False  
2          False  False  False  
3           True   True  False  
4          False  False  False  
..           ...    ...    ...  
339        False  False  False  
340        False  False  False  
341        False  False  False  
342        False  False  False  
343        False  False  False  

[344 rows x 9 columns]

In [79]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h2>FILTERING A COLUMN</h2>

In [80]:
penguins['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [81]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [82]:
penguins['species'] == 'Chinstrap'

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Name: species, Length: 344, dtype: bool

<h2>OPERATORS</h2>

In [83]:
penguins[penguins['species'] == 'Chinstrap']

rowid    species island  bill_length_mm  bill_depth_mm  \
276    277  Chinstrap  Dream            46.5           17.9   
277    278  Chinstrap  Dream            50.0           19.5   
278    279  Chinstrap  Dream            51.3           19.2   
279    280  Chinstrap  Dream            45.4           18.7   
280    281  Chinstrap  Dream            52.7           19.8   
..     ...        ...    ...             ...            ...   
339    340  Chinstrap  Dream            55.8           19.8   
340    341  Chinstrap  Dream            43.5           18.1   
341    342  Chinstrap  Dream            49.6           18.2   
342    343  Chinstrap  Dream            50.8           19.0   
343    344  Chinstrap  Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
276              192.0       3500.0  female  2007  
277              196.0       3900.0    male  2007  
278              193.0       3650.0    male  2007  
279              188.0       3525.0  female  2007  
280              197.0       3725.0    male  2007  
..                 ...          ...     ...   ...  
339              207.0       4000.0    male  2009  
340              202.0       3400.0  female  2009  
341              193.0       3775.0    male  2009  
342              210.0       4100.0    male  2009  
343              198.0       3775.0  female  2009  

[68 rows x 9 columns]

In [84]:
penguins[penguins['species'] != 'Chinstrap']

rowid species     island  bill_length_mm  bill_depth_mm  \
0        1  Adelie  Torgersen            39.1           18.7   
1        2  Adelie  Torgersen            39.5           17.4   
2        3  Adelie  Torgersen            40.3           18.0   
3        4  Adelie  Torgersen             NaN            NaN   
4        5  Adelie  Torgersen            36.7           19.3   
..     ...     ...        ...             ...            ...   
271    272  Gentoo     Biscoe             NaN            NaN   
272    273  Gentoo     Biscoe            46.8           14.3   
273    274  Gentoo     Biscoe            50.4           15.7   
274    275  Gentoo     Biscoe            45.2           14.8   
275    276  Gentoo     Biscoe            49.9           16.1   

     flipper_length_mm  body_mass_g     sex  year  
0                181.0       3750.0    male  2007  
1                186.0       3800.0  female  2007  
2                195.0       3250.0  female  2007  
3                  NaN          NaN     NaN  2007  
4                193.0       3450.0  female  2007  
..                 ...          ...     ...   ...  
271                NaN          NaN     NaN  2009  
272              215.0       4850.0  female  2009  
273              222.0       5750.0    male  2009  
274              212.0       5200.0  female  2009  
275              213.0       5400.0    male  2009  

[276 rows x 9 columns]

<h2>SEPARATING DATA FRAMES</h2>

<h3> CHINSTRAP DATA FRAMES </h3>


In [85]:
chinstrap_df = penguins[penguins['species'] == 'Chinstrap']
chinstrap_df

rowid    species island  bill_length_mm  bill_depth_mm  \
276    277  Chinstrap  Dream            46.5           17.9   
277    278  Chinstrap  Dream            50.0           19.5   
278    279  Chinstrap  Dream            51.3           19.2   
279    280  Chinstrap  Dream            45.4           18.7   
280    281  Chinstrap  Dream            52.7           19.8   
..     ...        ...    ...             ...            ...   
339    340  Chinstrap  Dream            55.8           19.8   
340    341  Chinstrap  Dream            43.5           18.1   
341    342  Chinstrap  Dream            49.6           18.2   
342    343  Chinstrap  Dream            50.8           19.0   
343    344  Chinstrap  Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
276              192.0       3500.0  female  2007  
277              196.0       3900.0    male  2007  
278              193.0       3650.0    male  2007  
279              188.0       3525.0  female  2007  
280              197.0       3725.0    male  2007  
..                 ...          ...     ...   ...  
339              207.0       4000.0    male  2009  
340              202.0       3400.0  female  2009  
341              193.0       3775.0    male  2009  
342              210.0       4100.0    male  2009  
343              198.0       3775.0  female  2009  

[68 rows x 9 columns]

<h3> ADELIE DATA FRAMES </h3>


In [86]:
adelie_df = penguins[penguins['species'] == 'Adelie']
adelie_df

rowid species     island  bill_length_mm  bill_depth_mm  \
0        1  Adelie  Torgersen            39.1           18.7   
1        2  Adelie  Torgersen            39.5           17.4   
2        3  Adelie  Torgersen            40.3           18.0   
3        4  Adelie  Torgersen             NaN            NaN   
4        5  Adelie  Torgersen            36.7           19.3   
..     ...     ...        ...             ...            ...   
147    148  Adelie      Dream            36.6           18.4   
148    149  Adelie      Dream            36.0           17.8   
149    150  Adelie      Dream            37.8           18.1   
150    151  Adelie      Dream            36.0           17.1   
151    152  Adelie      Dream            41.5           18.5   

     flipper_length_mm  body_mass_g     sex  year  
0                181.0       3750.0    male  2007  
1                186.0       3800.0  female  2007  
2                195.0       3250.0  female  2007  
3                  NaN          NaN     NaN  2007  
4                193.0       3450.0  female  2007  
..                 ...          ...     ...   ...  
147              184.0       3475.0  female  2009  
148              195.0       3450.0  female  2009  
149              193.0       3750.0    male  2009  
150              187.0       3700.0  female  2009  
151              201.0       4000.0    male  2009  

[152 rows x 9 columns]

<h3> GENTOO DATA FRAMES </h3>


In [87]:
gentoo_df = penguins[penguins['species'] == 'Gentoo']
gentoo_df

rowid species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152    153  Gentoo  Biscoe            46.1           13.2              211.0   
153    154  Gentoo  Biscoe            50.0           16.3              230.0   
154    155  Gentoo  Biscoe            48.7           14.1              210.0   
155    156  Gentoo  Biscoe            50.0           15.2              218.0   
156    157  Gentoo  Biscoe            47.6           14.5              215.0   
..     ...     ...     ...             ...            ...                ...   
271    272  Gentoo  Biscoe             NaN            NaN                NaN   
272    273  Gentoo  Biscoe            46.8           14.3              215.0   
273    274  Gentoo  Biscoe            50.4           15.7              222.0   
274    275  Gentoo  Biscoe            45.2           14.8              212.0   
275    276  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
271          NaN     NaN  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[124 rows x 9 columns]

<h3> HOW MANY CHINSTRAP IS WEIGHTING OVER 4000G </h3>

In [88]:
penguins['species'] == 'Chinstrap'

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Name: species, Length: 344, dtype: bool

In [89]:
penguins['body_mass_g'] >= 4000

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340    False
341    False
342     True
343    False
Name: body_mass_g, Length: 344, dtype: bool

In [90]:
chinstrap_heavy = penguins[(penguins['species'] == 'Chinstrap') & (penguins['body_mass_g'] <= 4000)]
chinstrap_heavy

rowid    species island  bill_length_mm  bill_depth_mm  \
276    277  Chinstrap  Dream            46.5           17.9   
277    278  Chinstrap  Dream            50.0           19.5   
278    279  Chinstrap  Dream            51.3           19.2   
279    280  Chinstrap  Dream            45.4           18.7   
280    281  Chinstrap  Dream            52.7           19.8   
281    282  Chinstrap  Dream            45.2           17.8   
282    283  Chinstrap  Dream            46.1           18.2   
283    284  Chinstrap  Dream            51.3           18.2   
285    286  Chinstrap  Dream            51.3           19.9   
286    287  Chinstrap  Dream            46.6           17.8   
287    288  Chinstrap  Dream            51.7           20.3   
288    289  Chinstrap  Dream            47.0           17.3   
290    291  Chinstrap  Dream            45.9           17.1   
292    293  Chinstrap  Dream            50.3           20.0   
293    294  Chinstrap  Dream            58.0           17.8   
294    295  Chinstrap  Dream            46.4           18.6   
296    297  Chinstrap  Dream            42.4           17.3   
297    298  Chinstrap  Dream            48.5           17.5   
298    299  Chinstrap  Dream            43.2           16.6   
299    300  Chinstrap  Dream            50.6           19.4   
300    301  Chinstrap  Dream            46.7           17.9   
302    303  Chinstrap  Dream            50.5           18.4   
303    304  Chinstrap  Dream            49.5           19.0   
304    305  Chinstrap  Dream            46.4           17.8   
306    307  Chinstrap  Dream            40.9           16.6   
308    309  Chinstrap  Dream            42.5           16.7   
310    311  Chinstrap  Dream            49.7           18.6   
311    312  Chinstrap  Dream            47.5           16.8   
312    313  Chinstrap  Dream            47.6           18.3   
314    315  Chinstrap  Dream            46.9           16.6   
316    317  Chinstrap  Dream            49.0           19.5   
317    318  Chinstrap  Dream            46.2           17.5   
318    319  Chinstrap  Dream            50.9           19.1   
319    320  Chinstrap  Dream            45.5           17.0   
320    321  Chinstrap  Dream            50.9           17.9   
322    323  Chinstrap  Dream            50.1           17.9   
324    325  Chinstrap  Dream            51.5           18.7   
325    326  Chinstrap  Dream            49.8           17.3   
326    327  Chinstrap  Dream            48.1           16.4   
327    328  Chinstrap  Dream            51.4           19.0   
328    329  Chinstrap  Dream            45.7           17.3   
330    331  Chinstrap  Dream            42.5           17.3   
331    332  Chinstrap  Dream            52.2           18.8   
332    333  Chinstrap  Dream            45.2           16.6   
334    335  Chinstrap  Dream            50.2           18.8   
335    336  Chinstrap  Dream            45.6           19.4   
336    337  Chinstrap  Dream            51.9           19.5   
337    338  Chinstrap  Dream            46.8           16.5   
338    339  Chinstrap  Dream            45.7           17.0   
339    340  Chinstrap  Dream            55.8           19.8   
340    341  Chinstrap  Dream            43.5           18.1   
341    342  Chinstrap  Dream            49.6           18.2   
343    344  Chinstrap  Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
276              192.0       3500.0  female  2007  
277              196.0       3900.0    male  2007  
278              193.0       3650.0    male  2007  
279              188.0       3525.0  female  2007  
280              197.0       3725.0    male  2007  
281              198.0       3950.0  female  2007  
282              178.0       3250.0  female  2007  
283              197.0       3750.0    male  2007  
285              198.0       3700.0    male  2007  
286              193.0       3800.0  female  2007  
287              194.0       3

In [91]:
chinstrap_heavy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 53 entries, 276 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   rowid              53 non-null     int64  
 1   species            53 non-null     object 
 2   island             53 non-null     object 
 3   bill_length_mm     53 non-null     float64
 4   bill_depth_mm      53 non-null     float64
 5   flipper_length_mm  53 non-null     float64
 6   body_mass_g        53 non-null     float64
 7   sex                53 non-null     object 
 8   year               53 non-null     int64  
dtypes: float64(4), int64(2), object(3)
memory usage: 4.1+ KB


<h2> FILLING UP MISSING VALUES </h2>

In [92]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [93]:
penguins['bill_length_mm'].isnull().sum()

np.int64(2)

In [94]:
penguins['bill_length_mm'].mean()

np.float64(43.9219298245614)

In [95]:
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())

In [96]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [97]:
penguins['bill_depth_mm'].mean()

np.float64(17.151169590643274)

In [98]:
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())

In [99]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [100]:
penguins['flipper_length_mm'].mean()

np.float64(200.91520467836258)

In [101]:
penguins['flipper_length_mm'] = penguins['flipper_length_mm'].fillna(penguins['flipper_length_mm'].mean())

In [102]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [103]:
penguins['body_mass_g'].mean()

np.float64(4201.754385964912)

In [104]:
penguins['body_mass_g'] = penguins['body_mass_g'].fillna(penguins['body_mass_g'].mean())

In [105]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           0
sex                  11
year                  0
dtype: int64

In [106]:
penguins['sex'].mode()

0    male
Name: sex, dtype: object

In [107]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])

In [108]:
penguins.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [109]:
#penguins_sex = penguins[penguins['sex'] == 'male']
#penguins_sex

<h2>CHANGING DATA TYPES</h2>

In [110]:
penguins

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen        39.10000       18.70000   
1        2     Adelie  Torgersen        39.50000       17.40000   
2        3     Adelie  Torgersen        40.30000       18.00000   
3        4     Adelie  Torgersen        43.92193       17.15117   
4        5     Adelie  Torgersen        36.70000       19.30000   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream        55.80000       19.80000   
340    341  Chinstrap      Dream        43.50000       18.10000   
341    342  Chinstrap      Dream        49.60000       18.20000   
342    343  Chinstrap      Dream        50.80000       19.00000   
343    344  Chinstrap      Dream        50.20000       18.70000   

     flipper_length_mm  body_mass_g     sex  year  
0           181.000000  3750.000000    male  2007  
1           186.000000  3800.000000  female  2007  
2           195.000000  3250.000000  female  2007  
3           200.915205  4201.754386    male  2007  
4           193.000000  3450.000000  female  2007  
..                 ...          ...     ...   ...  
339         207.000000  4000.000000    male  2009  
340         202.000000  3400.000000  female  2009  
341         193.000000  3775.000000    male  2009  
342         210.000000  4100.000000    male  2009  
343         198.000000  3775.000000  female  2009  

[344 rows x 9 columns]

In [111]:
penguins['species'] = penguins['species'].astype('category')
penguins['island'] = penguins['island'].astype('category')
penguins['sex'] = penguins['sex'].astype('category')
penguins['year'] = penguins['year'].astype('category')

In [112]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   rowid              344 non-null    int64   
 1   species            344 non-null    category
 2   island             344 non-null    category
 3   bill_length_mm     344 non-null    float64 
 4   bill_depth_mm      344 non-null    float64 
 5   flipper_length_mm  344 non-null    float64 
 6   body_mass_g        344 non-null    float64 
 7   sex                344 non-null    category
 8   year               344 non-null    category
dtypes: category(4), float64(4), int64(1)
memory usage: 15.4 KB


In [113]:
penguins.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [114]:
penguins.describe()

rowid  bill_length_mm  bill_depth_mm  flipper_length_mm  \
count  344.000000      344.000000     344.000000         344.000000   
mean   172.500000       43.921930      17.151170         200.915205   
std     99.448479        5.443643       1.969027          14.020657   
min      1.000000       32.100000      13.100000         172.000000   
25%     86.750000       39.275000      15.600000         190.000000   
50%    172.500000       44.250000      17.300000         197.000000   
75%    258.250000       48.500000      18.700000         213.000000   
max    344.000000       59.600000      21.500000         231.000000   

       body_mass_g  
count   344.000000  
mean   4201.754386  
std     799.613058  
min    2700.000000  
25%    3550.000000  
50%    4050.000000  
75%    4750.000000  
max    6300.000000

<h2>EXAMPLE QUESTION AND INSIGHTS</h2>

<h3>Which species has the heaviest body mass?</h3>

In [115]:
penguins['body_mass_g'].mean()

np.float64(4201.754385964912)

In [116]:
gentoo_df = penguins[penguins['species'] == 'Gentoo']

In [117]:
gentoo_df['body_mass_g'].mean()

np.float64(5068.965761177136)

In [118]:
chinstrap_df = penguins[penguins['species'] == 'Chinstrap']
chinstrap_df

rowid    species island  bill_length_mm  bill_depth_mm  \
276    277  Chinstrap  Dream            46.5           17.9   
277    278  Chinstrap  Dream            50.0           19.5   
278    279  Chinstrap  Dream            51.3           19.2   
279    280  Chinstrap  Dream            45.4           18.7   
280    281  Chinstrap  Dream            52.7           19.8   
..     ...        ...    ...             ...            ...   
339    340  Chinstrap  Dream            55.8           19.8   
340    341  Chinstrap  Dream            43.5           18.1   
341    342  Chinstrap  Dream            49.6           18.2   
342    343  Chinstrap  Dream            50.8           19.0   
343    344  Chinstrap  Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
276              192.0       3500.0  female  2007  
277              196.0       3900.0    male  2007  
278              193.0       3650.0    male  2007  
279              188.0       3525.0  female  2007  
280              197.0       3725.0    male  2007  
..                 ...          ...     ...   ...  
339              207.0       4000.0    male  2009  
340              202.0       3400.0  female  2009  
341              193.0       3775.0    male  2009  
342              210.0       4100.0    male  2009  
343              198.0       3775.0  female  2009  

[68 rows x 9 columns]

In [119]:
chinstrap_df['body_mass_g'].mean()

np.float64(3733.0882352941176)

In [120]:
adelie_df = penguins[penguins['species'] == 'Adelie']

In [121]:
adelie_df['body_mass_g'].mean()

np.float64(3703.95891043398)

<h4> INSIGHTS: Gentoo has the heaviest average body mass</h4>

<h2> HOW TO CREATE YOUR COLUMNS </h2>

In [122]:
penguins.columns

Index(['rowid', 'species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex', 'year'],
      dtype='object')

In [123]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   rowid              344 non-null    int64   
 1   species            344 non-null    category
 2   island             344 non-null    category
 3   bill_length_mm     344 non-null    float64 
 4   bill_depth_mm      344 non-null    float64 
 5   flipper_length_mm  344 non-null    float64 
 6   body_mass_g        344 non-null    float64 
 7   sex                344 non-null    category
 8   year               344 non-null    category
dtypes: category(4), float64(4), int64(1)
memory usage: 15.4 KB


In [124]:
penguins['mass_kg'] = np.round(penguins['body_mass_g'] /1000,1)

In [125]:
penguins.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
mass_kg              0
dtype: int64

<h1> Penguin Data Analysis: Questions, Code, and Insights</h1>

<h2>QUESTION 1</h2>

<h2>Which penguin is the most light?</h2>

In [126]:
penguins['mass_kg'].min()

np.float64(2.7)

In [127]:
penguins[penguins['mass_kg'] == 2.7]

rowid    species island  bill_length_mm  bill_depth_mm  \
314    315  Chinstrap  Dream            46.9           16.6   

     flipper_length_mm  body_mass_g     sex  year  mass_kg  
314              192.0       2700.0  female  2008      2.7

<h3>INSIGHTS: </h3>

The lightest penguin in the dataset is a female Chinstrap from Dream Island in 2008, weighing 2.7 kg. It has a smaller flipper length (192 mm) compared to the heaviest penguin and a slightly shorter bill length (46.9 mm).

<h2>QUESTION 2</h2>

<h2>How many female penguins belong to each species?</h2>

In [128]:
female_species_counts = penguins[penguins["sex"] == "female"]["species"].value_counts()
female_species_counts

species
Adelie       73
Gentoo       58
Chinstrap    34
Name: count, dtype: int64

<h3>INSIGHTS: </h3>

Adelie penguins have the highest number of females recorded, consistent with their overall greater presence in the dataset.

<h2>QUESTION #3</h2>

<h2>Which species has the widest range of flipper lengths (max - min)?</h2>

In [129]:
flipper_range = {
    species: penguins[penguins["species"] == species]["flipper_length_mm"].max() -
             penguins[penguins["species"] == species]["flipper_length_mm"].min()
    for species in penguins["species"].unique()
}
flipper_range


{'Adelie': np.float64(38.0),
 'Gentoo': np.float64(30.084795321637415),
 'Chinstrap': np.float64(34.0)}

<h3>INSIGHTS: </h3>

Adelie and Gentoo penguins show the same flipper length range, suggesting a broader variation in body size or age within these species.

<h2>QUESTION 4</h2>

<h2>What is the median mass of Adelie penguins?</h2>

In [130]:
adelie = penguins[penguins['species'] == 'Adelie']

In [131]:
adelie['mass_kg'].median()

np.float64(3.7)

<h3> INSIGHTS: </h3>


The median mass of Adelie penguins is 3.7 kg, indicating they are generally lighter than Gentoos. The median gives a more robust central value compared to the mean, especially in skewed data.


<h2>QUESTION 5</h2>

<h2>Which of the penguin is the max size?</h2>

In [132]:
penguins['mass_kg'].max()

np.float64(6.3)

<h3> INSIGHTS: </h3>

The heaviest penguin in the dataset is a male Gentoo from Biscoe Island in 2007, weighing 6.3 kg. It also has one of the longer flipper lengths (221 mm) and a relatively large bill length (49.2 mm).

<h2>QUESTION 6</h2>

<h2>Which penguin has the shortest flipper length?</h2>

In [133]:
penguins['flipper_length_mm'].min()

np.float64(172.0)

penguins[penguins['flipper_length_mm'] == 172.0]

<h3>INSIGHTS: </h3>

The penguin with the shortest flipper (172 mm) is a male Adelie from Torgersen in 2007, weighing 2.9 kg. Shorter flippers may suggest younger age or species-specific characteristics.

<h2>QUESTION 7</h2>

<h2>How many penguins were recorded each year?</h2>

In [134]:
year_counts = penguins['year'].value_counts()
year_counts

year
2009    120
2008    114
2007    110
Name: count, dtype: int64

<h3>INSIGHTS:</h3>

Data collection increased slightly each year from 2007 to 2009, showing consistent field research or expanding sample sizes over time.

<h2>QUESTION 8</h2>

<h2>What is the average flipper length of Gentoo penguins?</h2>

In [135]:
penguins[penguins['species']=='Gentoo']['flipper_length_mm'].mean()

np.float64(217.05576777966422)

<h3>INSIGHTS:</h3>

Gentoo penguins have the longest average flipper length (≈217 mm), aiding in fast swimming.

<h2>QUESTION 9</h2>

<h2>Which island has the most balanced sex distribution?</h2>

In [136]:
tbl = penguins.dropna(subset=['sex','island']).pivot_table(index='island', columns='sex', values='species', aggfunc='count', fill_value=0)
tbl

sex        female  male
island                 
Biscoe         80    88
Dream          61    63
Torgersen      24    28

<h3>INSIGHTS:</h3>

Biscoe Island tends to have the most balanced male-to-female ratio, which reduces sampling bias and helps in demographic studies.

<h2>QUESTION 10</h2>

<h2>Which species has the highest average bill length-to-depth ratio?</h2>

In [137]:
df = penguins.dropna(subset=['bill_length_mm','bill_depth_mm','species']).copy()


In [138]:
df['ratio'] = df['bill_length_mm'] / df['bill_depth_mm']

In [139]:
res = df.groupby('species')['ratio'].mean().sort_values(ascending=False)
res

species
Gentoo       3.170634
Chinstrap    2.653756
Adelie       2.122628
Name: ratio, dtype: float64

<h3>INSIGHTS:</h3>

Gentoo penguins usually show the highest bill length-to-depth ratio. A higher ratio means longer, slimmer bills, which may be adapted for specific feeding strategies.

<h2>QUESTION 11</h2>

<h2>Do male penguins show stronger correlation between body mass and flipper length compared to female penguins?</h2>

In [152]:
male_corr = penguins[penguins["sex"]=="male"]["flipper_length_mm"].corr(penguins[penguins["sex"]=="male"]["body_mass_g"])

In [153]:
female_corr = penguins[penguins["sex"]=="female"]["flipper_length_mm"].corr(penguins[penguins["sex"]=="female"]["body_mass_g"])

In [154]:
male_corr, female_corr

(np.float64(0.8628931650827829), np.float64(0.8848249756326794))

<h3>INSIGHTS:</h3>

Comparing genders shows if physical traits scale differently in males and females.

<h2>QUESTION 12</h2>

<h2>Is there a correlation between flipper length and body mass of penguins?</h2>

In [155]:
penguins["flipper_length_mm"].corr(penguins["body_mass_g"])

np.float64(0.8712017673060114)

<h3>INSIGHTS:</h3>

A strong positive correlation means bigger penguins tend to have longer flippers.

<h2>QUESTION 13</h2>

<h2>Which species has the highest correlation between bill length and body mass?</h2>

In [156]:
penguins.groupby("species").apply(lambda x: x["bill_length_mm"].corr(x["body_mass_g"]))

species
Adelie       0.553830
Chinstrap    0.513638
Gentoo       0.673636
dtype: float64

<h3>INSIGHTS:</h3>

This identifies which penguin species shows the strongest link between bill size and weight.